deepmd/pt/model/descriptor/repformer_layer.py

```diff
@@ -1,125 +1,324 @@
 # SPDX-License-Identifier: LGPL-3.0-or-later
 from typing import (
-    Callable,
     List,
+    Optional,
 )
 
 import torch
+import torch.nn as nn
 
-from deepmd.pt.model.network.network import (
-    SimpleLinear,
+from deepmd.pt.model.network.layernorm import (
+    LayerNorm,
+)
+from deepmd.pt.model.network.mlp import (
+    MLPLayer,
 )
 from deepmd.pt.utils import (
     env,
 )
+from deepmd.pt.utils.env import (
+    PRECISION_DICT,
+)
 from deepmd.pt.utils.utils import (
     ActivationFn,
+    to_numpy_array,
+    to_torch_tensor,
+)
+from deepmd.utils.version import (
+    check_version_compatibility,
 )
 
 
-def torch_linear(*args, **kwargs):
-    return torch.nn.Linear(
-        *args, **kwargs, dtype=env.GLOBAL_PT_FLOAT_PRECISION, device=env.DEVICE
+def get_residual(
+    _dim: int,
+    _scale: float,
+    _mode: str = "norm",
+    trainable: bool = True,
+    precision: str = "float64",
+) -> torch.Tensor:
+    r"""
+    Get residual tensor for one update vector.
+
+    Parameters
+    ----------
+    _dim : int
+        The dimension of the update vector.
+    _scale
+        The initial scale of the residual tensor. See `_mode` for details.
+    _mode
+        The mode of residual initialization for the residual tensor.
+        - "norm" (default): init residual using normal with `_scale` std.
+        - "const": init residual using element-wise constants of `_scale`.
+    trainable
+        Whether the residual tensor is trainable.
+    precision
+        The precision of the residual tensor.
+    """
+    residual = nn.Parameter(
+        data=torch.zeros(_dim, dtype=PRECISION_DICT[precision], device=env.DEVICE),
+        requires_grad=trainable,
     )
+    if _mode == "norm":
+        nn.init.normal_(residual.data, std=_scale)
+    elif _mode == "const":
+        nn.init.constant_(residual.data, val=_scale)
+    else:
+        raise RuntimeError(f"Unsupported initialization mode '{_mode}'!")
+    return residual
 
 
+# common ops
 def _make_nei_g1(
     g1_ext: torch.Tensor,
     nlist: torch.Tensor,
 ) -> torch.Tensor:
-    # nlist: nb x nloc x nnei
-    nb, nloc, nnei = nlist.shape
-    # g1_ext: nb x nall x ng1
+    """
+    Make neighbor-wise atomic invariant rep.
+
+    Parameters
+    ----------
+    g1_ext
+        Extended atomic invariant rep, with shape nf x nall x ng1.
+    nlist
+        Neighbor list, with shape nf x nloc x nnei.
+
+    Returns
+    -------
+    gg1: torch.Tensor
+        Neighbor-wise atomic invariant rep, with shape nf x nloc x nnei x ng1.
+
+    """
+    # nlist: nf x nloc x nnei
+    nf, nloc, nnei = nlist.shape
+    # g1_ext: nf x nall x ng1
     ng1 = g1_ext.shape[-1]
-    # index: nb x (nloc x nnei) x ng1
-    index = nlist.reshape(nb, nloc * nnei).unsqueeze(-1).expand(-1, -1, ng1)
-    # gg1  : nb x (nloc x nnei) x ng1
+    # index: nf x (nloc x nnei) x ng1
+    index = nlist.reshape(nf, nloc * nnei).unsqueeze(-1).expand(-1, -1, ng1)
+    # gg1  : nf x (nloc x nnei) x ng1
```

Comment:
not recommended to change "number of batch" to "number of frames" they basically mean the same thing...